File: online/LeetCodeCN/1~100/2.cpp
// https://leetcode-cn.com/problems/add-two-numbers/
// 模拟

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *head = new ListNode(0), *prev = head;

        int entry = 0;
        do {
            int left = l1 ? l1->val : 0;
            int right = l2 ? l2->val : 0;
            int sum = entry + left + right;

            entry = sum / 10;
            prev->next = new ListNode(sum % 10);
            prev = prev->next;

            l1 = l1 ? l1->next : nullptr;
            l2 = l2 ? l2->next : nullptr;
        } while (l1 || l2 || entry);

        return head->next;
    }
};

File: online/LeetCodeCN/1~100/2.ts
// 两数相加
// https://leetcode-cn.com/problems/add-two-numbers/

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    let head = new ListNode(), cur = head

    let entry = 0
    while (l1 != null || l2 != null || entry != 0) {
        let left = 0, right = 0
        if (l1) {
            left = l1.val
            l1 = l1.next
        }

        if (l2) {
            right = l2.val
            l2 = l2.next
        }

        const sum = left + right + entry
        const val = sum % 10;
        entry = Math.floor(sum / 10)

        cur.next = new ListNode(val)
        cur = cur.next
    }

    return head.next
};
